pygls/protocol.py

```diff
@@ -126,19 +133,21 @@ def deserialize_params(data, get_params_type):
 
         try:
             params_type = get_params_type(method)
+
             if params_type is None:
-                params_type = dict_to_object
-            elif params_type.__name__ == ExecuteCommandParams.__name__:
-                params = deserialize_command(params)
```

Comment:
As of this PR `deserialize_command` is a dead function... by *not* calling it seemed to make the type signature on [this method](https://github.com/openlawlibrary/pygls/blob/c495c16d5de20b364dea811af6d203f0117ed516/pygls/protocol.py#L675) actually reflect reality

Anyone know why it was being used before?

Reply:
So you're wondering if that was a workaround for something that still needs to be worked around, despite `deserialize_command` no longer existing?

Reply:
Yeah... it must've been added for a reason... but at the moment things appear more consistent without it

Reply:
@alcarney  Many thanks for all of your work! Not really sure if this is precisely what you're asking but as I have had to battle it out with this piece of code a while ago, here is  my rough understanding:



This seems like a temporary solution in order to reconcile 

- the desire to provide structured arguments rather than raw dictionaries to the various handlers for workspace/executeCommand 
- with the inability to implement proper validation without customizing the deserialization logic on a per command basis. (each method is expecting different kinds of arguments)

As a compromise, dictionaries inside the arguments are recursively translated into named tuples so as to  at least support attribute access like you can see 

https://github.com/perrinjerome/vscode-zc-buildout/blob/627b1ea6af66c4e4aad05603d8db231c577d6e5b/server/buildoutls/server.py#L95-L104

even if the objects do not (properly) match the corresponding type hints.


For my project needs, I have ended up implementing an alternative solution using [pydantic.validate_arguments](https://pydantic-docs.helpmanual.io/usage/validation_decorator/) which can wrap function to perform validation on its inputs based on the corresponding type hints. The argument list containing the raw data can then be passed directly without going through the whole named tuple conversion (I disable it by monkey patching `deserialize_command`). In my case I am also unpacking it for extra usability, for example in order to define a method with three parameters

```python
@server.custom_command("navigate_ast")
async def navigate_ast(
    server: PythonVoiceCodingPluginLanguageServer,
    command: StandardCommand,
    doc_uri: str,
    sel: Union[Range,Sequence[Range]] = [],
):
	...
```

where `StandardCommand` is a `pydantic.Model` defined in my project. My code looks roughly like this:

```python
def custom_command(
    self, command_name: str
) -> Callable[[F], Callable[["PythonVoiceCodingPluginLanguageServer", Any], Any]]:
    def wrapper(f: F):
        f = validate_arguments(config=dict(
            arbitrary_types_allowed=True))(f)

        async def function(server: PythonVoiceCodingPluginLanguageServer, args):
            return await f(server, *args)

        self.lsp.fm.command(command_name)(function)
        return f

    return wrapper
```

The whole thing needs to be polished, refined and rewritten for cattrs but I would be willing to help if there is interest to go down that way. An approach based on type hints might also go nicely with  #222 ?

Reply:
I certainly like the sound of de-serializing the arguments based on type hints!

@tombh since it sounds like this PR will likely land on a staging branch, would it make sense to remove the existing named tuple solution and have a follow on PR do something clever with type hints?

Reply:
I think so yes!

So you're saying to remove the existing tuple solution and move to the type hint solution in one PR? I mean there's no intermediate step? As in remove the tuple solution now, which would then require a replacement before merging into main.

And argument against this though is that we shouldn't put _too_ much in this release candidate. I mean it's better to focus on a minimum in order to get a RC public, rather than have too many new features that might block a minimal viable working lsprotocolised Pygls.